src/spine41.js

```diff
@@ -0,0 +1,669 @@
+
+        setWraps: function (uWrap, vWrap) {
+            this.pcTexture.addressU = TO_UV_WRAP_MODE[uWrap];
+            this.pcTexture.addressV = TO_UV_WRAP_MODE[vWrap];
+        },
+
+        getImage: function () {
+            return this._image;
+        },
+
+        dispose: function () {
+            this.pcTexture.destroy();
+        }
+    };
+
+    /**
+     * @name spine
+     * @namespace
+     * @description The Spine API.
+     */
+
+    /**
+     * @class
+     * @name pc.Spine
+     * @classdesc  A Spine animation object.
+     * @description Contains the skeleton and animation states as detailed in the Spine Runtime documentation.
+     * @param {pc.Application} app - The application that will manage this Spine object.
+     * @param {string} atlasData - Text data loaded from the atlas file.
+     * @param {object} skeletonData - JSON data loaded from the skeleton file.
+     * @param {object} textureData - Texture initialization data. An object where the key is the texture filename and the value is the pc.Texture resource.
+     * @property {spine.Skeleton} skeleton The Skeleton object.
+     * @property {spine.AnimationState} state The first AnimationState object. There is always one AnimationState.
+     * @property {spine.AnimationState[]} states A list of all AnimationState objects.
+     * @property {number} priority An integer value which determines when the spine mesh is rendered relative to other Spine meshes. Lower numbers are rendered first.
+     * @property {boolean} autoUpdate Determines whether the Spine object calls skeleton.updateWorldTransform in the update loop. Default is true.
+     */
+    var Spine = function (app, atlasData, skeletonData, textureData) {
+        this._app = app;
+
+        this._position = new pc.Vec3();
+
+        // create the atlas
+        var atlas = new spine.TextureAtlas(atlasData);
+
+        // load required texture for each page
+        for (const page of atlas.pages) {
+            // todo: handle aliases
+
+            // set the textures
+            page.setTexture(new SpineTextureWrapper(textureData[page.name]));
+        }
+
+        // Create a AtlasAttachmentLoader that resolves region, mesh, boundingbox and path attachments
+        var atlasLoader = new spine.AtlasAttachmentLoader(atlas);
+
+        // Create a SkeletonJson instance for parsing the .json file.
+        var skeletonJson = new spine.SkeletonJson(atlasLoader);
+
+        // Set the scale to apply during parsing, parse the file, and create a new skeleton.
+        skeletonJson.scale *= 0.01;
+        var _skeletonData = skeletonJson.readSkeletonData(skeletonData);
+
+        // Compatibility queries
+        this.skeletonVersion = _skeletonData.version;
+        this._spine_3_6_0 = versionCompare(this.skeletonVersion, "3.6.0") <= 0;
+        this._spine_3_7_99 = versionCompare(this.skeletonVersion, "3.7.99") <= 0;
+        this._spine_3_7_99 = versionCompare(this.skeletonVersion, "4.1.23") <= 0;
```

Comment:
This looks wrong. 😄

Reply:
Possibly, the intent is to be using the stuff from that flag "_spine_3_7_9" with first version of the 4.1 file format that became stable.  I think it goes up to "4.1.36".

I'm going for clarity on the exact changes required. The example source they provide for the engines I know well supports much more / checks other things, etc.

Reply:
Sorry, I'm not quite sure I follow. Line 158 simply overwrites what was written on line 157.

Reply:
Your right.  It's because of a deeper bug with the `versionCompare` function.  I'll update.

Short story, it does not support versions like "4.1.23-beta".

Issue:

Spine files can have non numeric version values in compliance with `semver`.  The json file in the 4.1 example has "4.1.23-beta".  `versionCompare` will return `nan` when it's not getting numbers.  The lexical compare option in `versionCompare` is not fully compliant with `semver` so that doesn't sort the issue.

Resolution: 

- switch to `semver` library parsing like the Spine lib.

Details:

The current `versionCompare` is a copy / paste form this stackoverflow post.

https://stackoverflow.com/questions/6832596/how-can-i-compare-software-version-number-using-javascript-only-numbers

But it has a few issues as noted in this gist:

https://gist.github.com/TheDistantSea/8021359

Reply:
Resolved in: [0c72c83](https://github.com/playcanvas/playcanvas-spine/pull/73/commits/0c72c83ea0279190593704a389a4d9aef37cc6a6)